docs/tutorial/codecs.txt

```diff
@@ -0,0 +1,433 @@
+
+       public function decode($value): DateTimeImmutable
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           $utc = $value instanceof MongoDB\BSON\UTCDateTime
+               ? $value
+               : $value->get('utc');
+
+           if (! $utc instanceof UTCDateTime) {
+               throw UnsupportedValueException::invalidDecodableValue($utc);
+           }
+
+           $dateTime = $utc->toDateTime();
+
+           if ($value instanceof MongoDB\BSON\Document && $value->has('tz')) {
+               $dateTime->setTimeZone(new DateTimeZone($value->get('tz')));
+           }
+
+           return DateTimeImmutable::fromMutable($dateTime);
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           return MongoDB\BSON\Document::fromPHP([
+               'utc' => new UTCDateTime($value),
+               'tz' => $value->getTimezone()->getName(),
+           ]);
+       }
+   }
+
+This codec can now be leveraged by other codecs encode and decode dates. Let's return to the previous example and add a
+``createdAt`` field that contains the creation date. The modified encode and decode methods would look like this (other
+code omitted for brevity):
+
+.. code-block:: php
+
+   <?php
+
+   final class PersonCodec implements MongoDB\Codec\DocumentCodec
+   {
+       public function __construct(
+           private readonly DateTimeCodec $dateTimeCodec = new DateTimeCodec(),
+       ) {}
+
+       // Other code omitted for brevity
+       public function decode($value): Person
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           $person = new Person($value->get('name'));
+           $person->id = $value->get('_id');
+           $person->createdAt = $this->dateTimeCodec->decode($value->get('createdAt'));
+
+           return $person;
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           return MongoDB\BSON\Document::fromPHP([
+               '_id' => $value->id,
+               'name' => $value->name,
+               'createdAt' => $this->dateTimeCodec->encode($value->createdAt),
+           ]);
+       }
+   }
+
+Handling embedded documents
+---------------------------
+
+The previous example showed how to handle a single document. However, sometimes you want to handle fields that contain
+embedded documents. To show this, let's create an address document that we'll be embedding into our ``Person`` document.
+To ensure consistency, we're going to make this a readonly class:
```

Comment:
```suggestion
To ensure consistency, we're going to make this a read-only class:
```

I realize the PHP keyword is `readonly`, but I think this reads better in English.